[Spring 7주차 과제] 인가(Authorization) 구현하기

developerOlive opened the pull request:
**과제 목표**
지금은 로그인을 했다고 하더라도, 다른 사람의 정보를 내 마음대로 수정할 수 있습니다. 
내 정보는 오직 나만 수정할 수 있어야 합니다. 
오직 나만이 내 정보를 수정할 수 있도록 Spring Security를 이용해서 구현해 주세요.

현재 사용자의 비밀번호를 평문으로 저장하고 있습니다.
누군가 외부에서 침투를 하거나 혹은 내부에서 비밀번호를 탈취를 하더라도 원래 비밀번호를 알아낼 수 없도록 비밀번호를 암호화하여 저장해 주세요.

---------------------------------------------------------------------------------------

📍로그인이 필요없는 API
- [x] 로그인 - POST /session
- [x] 회원 생성하기 - POST /users
- [x] 고양이 장난감 목록 얻기 - GET /products
- [x] 고양이 장난감 상세 조회하기 - GET /products/{id}
---------------------------------------------------------------------------------------

📍로그인이 필요한 API
- [x] 고양이 장난감 등록하기 - POST /products
- [x] 고양이 장난감 수정하기 - PATCH /products/{id}
- [x] 고양이 장난감 삭제하기 - DELETE /products/{id}
- [x] 회원 수정하기 - POST /users/{id}
- [x] 회원 삭제하기 - DELETE /users/{id}
---------------------------------------------------------------------------------------

📍요구 사항
- [x] 테스트 커버리지 100%를 달성해야 합니다.
- [x] 모든 API 테스트를 통과해야 합니다.
- [x] 모든 E2E 테스트를 통과해야 합니다.

developerOlive committed:
Feat : .gitIgnore 세팅

developerOlive committed:
Feat : Spring security 사용을 위해 build.gradle에 의존성 추가

developerOlive committed:
Feat : SecurityJavaConfig 생성 후, CSRF 비활성화를 위한 코드 추가

developerOlive committed:
Feat : 필터를 사용하기 위해 authenticationFilter 생성

- JwtAuthenticationFilter을 사용하기 위해서 authenticationManager()을 매개변수로 받음
(authenticationManager : Spring Security의 필터들이 인증을 수행하는 방법에 대한 명세를 정의해 놓은 인터페이스)

developerOlive committed:
Feat : JwtAuthenticationFilter에 doFilterInternal 생성

developerOlive committed:
Feat :  doFilterInternal 메소드 내 인증을 위한 코드 추가

JwtAuthenticationFilter - authenticationService을 사용하기 위해 필드에 선언 후 @Autowired 추가, 생성자에도 추가
SecurityJavaConfig - authenticationService을 사용하기 위해 필드에 선언 후 @Autowired 추가, 생성자에도 추가

developerOlive committed:
Feat : filterWithPathAndMethod 추가

developerOlive committed:
Feat : SecurityJavaConfig - .addFilterBefore 추가

developerOlive committed:
Feat : AuthenticationErrorFilter 생성

developerOlive committed:
Feat : 인터셉터 관련 코드 삭제

developerOlive commented:
안녕하세요, 트레이너님.
어느덧 7주차 입니다!

일단 첫번째 강의 수강 후 공부한 코드 커밋합니다.
여러 번 봤는데도 많이 어렵네요... 
중간중간 자료도 찾아봐야 이해가 되서 시간이 꽤 많이 걸릴 것 같아요 😭

hannut91 commented:
Spring Security가 상당히 복잡해서, 모든 것을 공부하고 이해하고 하기에는 벅찹니다 ㅎㅎ 최소한의 기능으로 기능을 구현해보면서 내가 필요한 기능을 사용하는 방법을 익힌다고 생각해보시면 좋을 것 같습니다.

developerOlive committed:
Refactor : JwtAuthenticationFilter - Path Method 분기하는 코드를 지우기 위한 리팩토링

developerOlive committed:
Feat : User 인증을 처리하기 위해 UserAuthentication 생성

developerOlive committed:
Refactor : SecurityJavaConfig에 .exceptionHandling() 관련 코드 추가

developerOlive committed:
Feat : ProductControllerTest - 엑세스 토큰 없이 삭제하는 상황에 대한 테스트 코드 추가

developerOlive committed:
Refactor : ProductController - 각 메소드 매개변수로 Authentication을 받도록 수정

developerOlive committed:
Refactor : AuthenticationService - login 메서드 리턴값 매개변수 수정

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/ProductController.java:
```diff
@@ -17,6 +20,7 @@
 @RestController
 @RequestMapping("/products")
 @CrossOrigin
+@Slf4j
```
오 이건 어디에 사용되는거에요?

developerOlive replied:
System.out.println 대신에 로그를 찍을 수 있도록 도와주는 에노테이션입니다! 

![image](https://user-images.githubusercontent.com/67456294/110769558-a5e93d00-829b-11eb-9e3e-bfc5d00537a8.png)

아래와 같이 콘솔창에 로그를 찍어줍니다.

![image](https://user-images.githubusercontent.com/67456294/110769498-936f0380-829b-11eb-853b-4b8638a4eb30.png)

developerOlive committed:
Refactor : UserControllerTest - 헤더에 토큰이 필요한 테스트에 토큰 추가

developerOlive committed:
Refactor : 자신의 계정으로만 수정가능하도록 처리 및 관련 테스트 코드 수정

developerOlive committed:
Feat : 테스트 커버리지 항목에서 lombok builder 관련 코드를 제외시키기 위한 에노테이션 추가

developerOlive committed:
Feat :  h2-console을 사용할 수 없는 에러 해결을 위한 코드 추가

developerOlive committed:
Feat : 사용되지 않는 ExceptionHandler 삭제

developerOlive committed:
Feat : 암호화 코드 작성 및 관련 테스트 코드에 암호화 매개변수 추가

정보 보호를 위해 비밀번호를 사용자가 입력한 그대로 저장하지 않고, 암호화 하여 저장하도록 수정.
관련된 테스트 코드에 비밀번호를 암호화 할 수 있도록 passwordEnoder 매개변수 추가.

developerOlive committed:
Feat : UserControllerSerivceTest 테스트 코드 추가

- 이미 존재하는 이메일로 가입하려는 경우
- 유효하지 않은 토큰이 주어진 경우

developerOlive committed:
Feat : UserServiceTest - 인가되지 않은 상황에서 수정을 진행하려는 경우 테스트 코드 추가

developerOlive committed:
Refactor : 서버 구동이 안되는 에러 처리

developerOlive committed:
Feat : ProductControllerTest - AccessDeniedException이 발생할 경우의 테스트 코드 추가

developerOlive committed:
Refactor :  updateUserWithAccessDeniedToken - 검증하는 테스트 코드 추가

developerOlive committed:
Feat : 테스트 커버리지에서 테스트하지 않을 코드 애노테이션을 활용하여 제외

developerOlive committed:
ReTest

developerOlive committed:
ReTest

developerOlive committed:
Refactor : 불필요한 import 삭제

developerOlive committed:
Refactor : 테스트 독립성 처리

developerOlive committed:
Retest

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -28,17 +33,30 @@ public User registerUser(UserRegistrationData registrationData) {
         user.changeWith(source);
 
         return user;
+
```
여기 빈 줄은 의도가 있는걸까요?

developerOlive replied:
없습니다! 삭제할게요 ㅎㅎ

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -28,11 +31,13 @@ UserResultData create(@RequestBody @Valid UserRegistrationData registrationData)
+            Authentication authentication
     ) {
-        User user = userService.updateUser(id, modificationData);
+        User user = userService.updateUser(authentication, id, modificationData);
```
인증처리하는 것은 Spring Security에서 모두 처리되고, Controller에는 인증된 정보만 넘겨줍니다.
그리고 UserService에는 인증이라는 정보가 계속 전달되지 않았으면 좋겠습니다. 왜냐하면 서비스가 도메인 안쪽으로 의존하는 것이 아니라, Controller 바깥으로 의존하게 되기 때문입니다.

의존성은 바깥에서 안쪽으로 의존해야 합니다.

![image](https://user-images.githubusercontent.com/14071105/110977770-e2a15b00-83a5-11eb-871d-368569a8479a.png)

See also
  - https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html

developerOlive replied:
앗 그렇군요! 한 번 살펴보겠습니다!

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -28,11 +34,29 @@
+
+    private Mapper mapper;
+
+    private static final String USER2_EMAIL_ADDRESS = "test2@test.com";
+    private static final Long NOT_EXIST_USER_ID = 100L;
+    private static final Long USER1_ID = 1L;
+    private static final Long USER2_ID = 2L;
+    private static final String USER1_NAME = "Tester";
+    private static final String USER2_NAME = "Tester2";
+    private static final String USER1_PASSWORD = "test";
+    private static final String USER2_PASSWORD = "test2";
+    private static final String MODIFIED_NAME = "TEST";
+    private static final String MODIFIED_PASSWORD = "TEST";
```
user1, user2를 만들어놓고 사용할 수 있나요? 예를들어서

```java
User user1 = User.builder()
                           .id(1L)
                           .email("user1@email.com");
                           
User user1 = User.builder()
                           .id(2L)
                           .email("user2@email.com");
```

위와 같이 만들어놓고 테스트에서 다음과 같이 사용하는 것은 어떨까요?

```java
given(userRepository.findByIdAndDeletedIsFalse(user1.getId()))
                .willReturn(Optional.of(user1));
```

developerOlive committed:
Refactor : 테스트 통과를 위해 JwtUtil - encode 코드 내용 수정

developerOlive committed:
ReTest

developerOlive committed:
ReTest

developerOlive committed:
ReTest

developerOlive commented:
push하는 과정에서 깃허브 서버 오류로 인해, 
원격 저장소에 push한 내용이 업로드 되지 않아 ReTest를 여러 번 하게 되었습니다. 
참고 부탁 드립니다!

developerOlive committed:
docs : AuthenticationService JavaDoc 추가

developerOlive committed:
docs : ProductService, UserService JavaDoc 작성

developerOlive committed:
Refactor : SecurityJavaConfig - 생성자 주입 부분 코드 수정

developerOlive commented:
트레이너님, 이번주는 주제가 난이도가 있는 만큼 디버깅 하거나 코드를 이해하는데 생각보다 많은 시간이 소요되네요..

테스트는 통과했지만 아직 흐름이 이해되지 않는 부분이 꽤 있어서 
javaDoc을 작성하면서 각 클래스가 어떤 일을 하는지,
흐름이 어떻게 되는지 복습하려고 합니다.

구현하고 에러 해결하느라 커밋이나 피드백이 다른때보다 많지는 않았는데
난이도가 이제까지 학습했던 것 중에서 제일 높았던 것 같아요.🤧

https://github.com/CodeSoom/spring-week7-assignment-1/pull/7#discussion_r593350399
이 피드백대로 수정하려면 건드려야 될 곳이 좀 있어보이네요.. ㅎㅎ ㅠㅠ

일단 과제풀이 영상 보겠습니다!

hannut91 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -5,7 +5,10 @@
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
 
+@lombok.Generated
```
이건 아마 Lombok에서 자동으로 만들어주는 코드들에 대해서 테스트하지 않도록 하기 위해서 추가한걸로 기억하는데, 여기서는 Lombok이 사용되고 있지 않자나요. 그래도 여기에 이렇게 쓰면 테스트에서 무시가 되나요?

developerOlive replied:
아래 이미지는 @lombok.Generated 를 지웠을 때의 상황입니다.

![image](https://user-images.githubusercontent.com/67456294/111068135-d4b91a80-850a-11eb-8628-fca17728b823.png)

저도 이 부분을 테스트하지 않기 위해 혹시나 하고 써봤는데 테스트 무시가 되더라고요.

hannut91 replied:
뭔가 편법같지만 되는군요 흠..

developerOlive committed:
Feat : SecurityJavaConfig - JavaDoc 및 코드설명 추가

developerOlive committed:
Feat : AuthenticationErrorFilter - JavaDoc 및 코드설명 추가

developerOlive committed:
Feat : JwtAuthenticationFilter - JavaDoc 및 코드설명 추가

hannut91 commented:
와 ㅎㅎ 회고를 보니 거의 학습한 모든 과정이 기록되어 있네요 ㅎㅎ 각 내용들이 블로그글 하나로 써도 될 정도로 엄청나네요
마지막 자기선언이 `살아남자`라니 ㅎㅎ 역시 꾸준히 열심히 하시는 분들이 끝까지 살아남으시는 것 같습니다
이번주 고생 많으셨습니다.
